src/winrtble/peripheral.rs

```diff
@@ -395,27 +395,48 @@ impl ApiPeripheral for Peripheral {
             let gatt_services = device.discover_services().await?;
             for service in &gatt_services {
                 let uuid = utils::to_uuid(&service.Uuid().unwrap());
-                match BLEDevice::get_characteristics(&service).await {
-                    Ok(characteristics) => {
-                        let characteristics = characteristics
-                            .into_iter()
-                            .map(|gatt_characteristic| {
-                                let characteristic = BLECharacteristic::new(gatt_characteristic);
-                                (characteristic.uuid(), characteristic)
-                            })
-                            .collect();
-                        self.shared.ble_services.insert(
-                            uuid,
-                            BLEService {
+                let mut retry = 0;
+                // Try to get the characteristics for a service (3 attempts).
+                // For some reason, get_characteristics() sometimes returns an
```

Comment:
get_characteristics uses `GattDeviceService::GetCharacteristicsAsync`, which may take values from a system-wide cache. Could this be solved by switching to [GetCharacteristicsWithCacheModeAsync(BluetoothCacheMode::Uncached)](https://docs.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.genericattributeprofile.gattdeviceservice.getcharacteristicsasync?view=winrt-22000#windows-devices-bluetooth-genericattributeprofile-gattdeviceservice-getcharacteristicsasync(windows-devices-bluetooth-bluetoothcachemode))? Similar change was recently made to `get_gatt_services`. It might of course have some performance impact, so one alternative could be to first use the cache and if 0 characteristics are found, try uncached.

Reply:
tl;dr: Using Uncached makes the problem worse, but I think I can improve the logic by moving the retry loop down a level where it can be triggered by the error state.



That sounds plausible, and was worth trying, but I'm actually seeing the issues even more consistently when using uncached vs cached.

What I didn't spot earlier was that get_characteristics() is a wrapper, and that the error handling happens at that next level: turns out that the problem is random failures of "access denied": `get_status Ok(GattCommunicationStatus(3))`. In hindsight, this is where the retry loop should be: where the error is known and not just inferred.

I bumped the log levels on the output in that method to make the issue clearer (this is without the proposed retry loop, using cached characteristics):
```
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(3)    
ERROR device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: get_status Ok(GattCommunicationStatus(3))    
ERROR device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: get_status Ok(GattCommunicationStatus(3))    
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(5)    
ERROR device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: buttplug::device: Failed to create device implementation with protocol satisfyer: ButtplugDeviceError(InvalidEndpoint(Command))
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(3)    
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(8)    
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(5)    
INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: buttplug::device: Found Buttplug Device Curve match protocol magic-motion-2 address=PeripheralId(FF:FF:20:0F:4F:EE)
```
The code scans and iterates over the services and their characteristics (twice, once per protocol match attempt), you should be able to see the 2 characteristic scans that fail the first time, and succeed the second time.

I almost wonder whether the characteristic scan happens under the covers regardless of whether you ask it to, and if you do ask it to scan for characteristics whilst the stack is already doing so, you get the access denied error?

Reply:
Quick update:
* Putting the retry loop in `get_characteristics()` did not work: if a call to `GetCharacteristicsWithCacheModeAsync()` failed with Access Denied, it failed consistently for the passed in service object.
* The above behaviour was seemingly made worse when using Uncached (I don't have hard numbers for that, I just perceived that I saw the errors more often
* Putting the retry loop back into `discover_services()`, I don't see any Access Denied errors no matter how many times I run the code, without the retry loop ever looping.

Either there's a weird timing error in `discover_services()` around the services or there's some sort of lifetime bug in there...

Reply:
I'm leaning towards this being a lifetime issue, but my Rust isn't good enough to understand how that can be.

In the two following cases I see the Access Denied errors intermittently:
````
                let uuid = utils::to_uuid(&service.Uuid().unwrap());
                info!("checking service {:?}", uuid);
                    match BLEDevice::get_characteristics(&service).await {
```
```
                let uuid = utils::to_uuid(&service.Uuid().unwrap());
                Delay::new(Duration::from_millis(200)).await;
                    match BLEDevice::get_characteristics(&service).await {
```
In these cases I can't reproduce the error:
```
                let uuid = utils::to_uuid(&service.Uuid().unwrap());
                if !self.shared.ble_services.contains_key(&uuid) {
                    match BLEDevice::get_characteristics(&service).await {
```
```
                let uuid = utils::to_uuid(&service.Uuid().unwrap());
                while !self.shared.ble_services.contains_key(&uuid) {
```

Some more debugging later, and it appears that multiple calls to the service object to ask for characteristics seems to be leading to the Access Denied error (each after a call to `device.discover_services()`), in other words calling `Peripheral::discover_services()` multiple times on the same device seems to cause the intermittent issue:

```
2022-02-11T08:17:37.593648Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: buttplug::device::configuration_manager: Found protocol "satisfyer" for specifier BluetoothLE(BluetoothLESpecifier { names: {"Curve"}, advertised_services: {0000180a-0000-1000-8000-00805f9b34fb}, required_endpoints: {}, services: {} }).
2022-02-11T08:17:37.594628Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: buttplug::device::configuration_manager: Found protocol "magic-motion-2" for specifier BluetoothLE(BluetoothLESpecifier { names: {"Curve"}, advertised_services: {0000180a-0000-1000-8000-00805f9b34fb}, required_endpoints: {}, services: {} }).
2022-02-11T08:17:40.145473Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? false    
2022-02-11T08:17:42.178288Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
2022-02-11T08:17:42.183244Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? false    
2022-02-11T08:17:43.526164Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(3)    
2022-02-11T08:17:43.547065Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? false    
2022-02-11T08:17:44.010169Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
2022-02-11T08:17:44.011258Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? false    
2022-02-11T08:17:45.160568Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(8)    
2022-02-11T08:17:45.165422Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? false    
2022-02-11T08:17:45.553958Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
2022-02-11T08:17:45.559895Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? false    
2022-02-11T08:17:46.361128Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(5)    
2022-02-11T08:17:46.362748Z ERROR device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: buttplug::device: Failed to create device implementation with protocol satisfyer: ButtplugDeviceError(InvalidEndpoint(Command))
2022-02-11T08:17:46.750997Z  INFO device creation{name= address=PeripheralId(FF:FF:20:0F:4F:EE)}: buttplug::device: No protocol were matched
2022-02-11T08:17:48.024849Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:48.080143Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
2022-02-11T08:17:48.088974Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:48.249915Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(3)    
2022-02-11T08:17:48.251956Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:48.257757Z ERROR device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: get_characteristics_async Other("Ok(GattCommunicationStatus(3))")    
2022-02-11T08:17:48.257880Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:48.362119Z ERROR device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: get_characteristics_async Other("Ok(GattCommunicationStatus(3))")    
2022-02-11T08:17:48.362251Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:48.589752Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
2022-02-11T08:17:48.631947Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:48.788362Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(5)    
2022-02-11T08:17:48.806171Z ERROR device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: buttplug::device: Failed to create device implementation with protocol satisfyer: ButtplugDeviceError(InvalidEndpoint(Command))
2022-02-11T08:17:49.668410Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:49.695222Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
2022-02-11T08:17:50.050805Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:50.326359Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(3)    
2022-02-11T08:17:50.333477Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:50.609863Z ERROR device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: get_characteristics_async Other("Ok(GattCommunicationStatus(3))")    
2022-02-11T08:17:50.610377Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:50.804072Z ERROR device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: get_characteristics_async Other("Ok(GattCommunicationStatus(3))")    
2022-02-11T08:17:50.804848Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:51.018986Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
2022-02-11T08:17:51.024855Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::peripheral: Already got service? true    
2022-02-11T08:17:51.186173Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: btleplug::winrtble::ble::device: characteristics Ok(5)    
2022-02-11T08:17:51.300806Z  INFO device creation{name=Curve address=PeripheralId(FF:FF:20:0F:4F:EE)}: buttplug::device: Found Buttplug Device Curve match protocol magic-motion-2 address=PeripheralId(FF:FF:20:0F:4F:EE)
```
This behaviour seems to be the same regardless of the use of cached/uncached.